### PR TITLE
Merge upstream changes up to commit  `d3456309`

### PR DESCRIPTION
**Pull request by illia-li**
Original commit list:
**1. https://github.com/apache/cassandra-gocql-driver/commit/3a475c75aefc97b2961bae3a2d53b3e87c9274d0**
Message:
```
CASSGO-44: keep nil slices in MapScan
Previously if you read a NULL column as a varchar using MapScan it
would store the value []byte{} in the map rather than []byte() which
made it impossible to know if the value was NULL or not.

Patch by James Hartig; reviewed by João Reis for CASSGO-44
```
Status: `APPLIED`

**2. https://github.com/apache/cassandra-gocql-driver/commit/c63468d9ce9dca353f55a3371aa97a30eedd7d4b**
Message:
```
CASSGO-42: don't panic if no applied column is returned
I also added checks against MapScan failing which could also trigger this bug.

Several integration tests were added to validate various edge cases.

Patch by James Hartig; reviewed by João Reis for CASSGO-42
```
Status: `APPLIED`

**3. https://github.com/apache/cassandra-gocql-driver/commit/91cbf12722ef7a53cfd0f0e94fe86233426dd34f**
Message:
```
skip metadata only if prepared result included metadata
If the prepare result doesn't include metadata we shouldn't skip metadata
when executing the query. Fixes CASSGO-40.

Patch by James Hartig; reviewed by João Reis for CASSGO-40
```
Status: `APPLIED`

**4. https://github.com/apache/cassandra-gocql-driver/commit/16366d4eff6a21c769203299330f6109de9f2401**
Message:
```
Improve error messages for marshalling
Enhancement for marshal/unmarshal error messages.
Errors were updated by the supported types list.

patch by Oleksandr Luzhniy; reviewed by João Reis, James Hartig, Bohdan Siryk, for [CASSGO-38](https://issues.apache.org/jira/browse/CASSGO-38)
```
Status: `APPLIED`

**5. https://github.com/apache/cassandra-gocql-driver/commit/d3456309946b01151c8d0329c2e11c1d192001ff**
Message:
```
CASSGO-2 Marshal big int returns variable length slice.
The marshalBigInt return 8 bytes slice in all cases except for big.Int,
which returns a variable length slice, but should be 8 bytes slice as well.

patch by Mykyta Oleksiienko; reviewed by João Reis, James Harting for CASSGO-2
```
Status: `APPLIED`

### CODE REVIEW

File: conn.go [hunk 1]
```diff
@@ -1484,7 +1484,8 @@ func (c *Conn) executeQuery(ctx context.Context, qry *Query) (iter *Iter) {
 			}
 		}
 
-		params.skipMeta = !(c.session.cfg.DisableSkipMetadata || qry.disableSkipMetadata)
+		// if the metadata was not present in the response then we should not skip it
+		params.skipMeta = !(c.session.cfg.DisableSkipMetadata || qry.disableSkipMetadata) && info != nil && info.response.flags&flagNoMetaData == 0
```
Comment (illia-li) [inline conn.go:1488]: Please look at this change. 
In my opinion, it should look like this:
```
params.skipMeta = !(c.session.cfg.DisableSkipMetadata || qry.disableSkipMetadata) && info != nil && info.response.flags&flagNoMetaData != 0
```

**Comment by dkropachev** [comment]
> **4. [apache@16366d4](https://github.com/apache/cassandra-gocql-driver/commit/16366d4eff6a21c769203299330f6109de9f2401)** Message:
> 
> ```
> Improve error messages for marshalling
> Enhancement for marshal/unmarshal error messages.
> Errors were updated by the supported types list.
> 
> patch by Oleksandr Luzhniy; reviewed by João Reis, James Hartig, Bohdan Siryk, for [CASSGO-38](https://issues.apache.org/jira/browse/CASSGO-38)
> ```
> 
> Status: `APPLIED`

Can you please double check, it looks like this one is completely dropped.

File: conn.go (hunk 1 above)
Comment (dkropachev) [inline conn.go:1488]: Let's make it cover case when `flagNoMetaData` is not set, but no response columns provided.
```suggestion
		// if the metadata was not present in the response then we should not skip it
		params.skipMeta = !(c.session.cfg.DisableSkipMetadata || qry.disableSkipMetadata) && len(info.response.columns) != 0
```

Comment (illia-li) [reply conn.go]: Done

File: conn.go [hunk 2]
```diff
@@ -1594,8 +1595,6 @@ func (c *Conn) executeQuery(ctx context.Context, qry *Query) (iter *Iter) {
 			} else {
 				return &Iter{framer: framer, err: errors.New("gocql: did not receive metadata but prepared info is nil")}
 			}
-		} else {
-			iter.meta = x.meta
```
Comment (illia-li) [inline conn.go:1598]: Also double metadata copying was removed.

**Review by dkropachev — APPROVED** [review]
